tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,587 @@
+		return;
+	}
+
+	server->accept = accept;
+	server->psm = 0;
+
+	if (bt_l2cap_server_register(server) < 0) {
+		FAIL("Failed to get free server");
+		return;
+	}
+
+	LOG_DBG("L2CAP server registered, PSM:0x%X", server->psm);
+}
+
+static void connected(struct bt_conn *conn, uint8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		FAIL("Failed to connect to %s (%u)", addr, conn_err);
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
+	LOG_DBG("%s", addr);
+
+	SET_FLAG(is_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_DBG("%s (reason 0x%02x)", addr, reason);
+
+	if (default_conn != conn) {
+		FAIL("Conn mismatch disconnect %s %s)", default_conn, conn);
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+	UNSET_FLAG(is_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void send_sdu(int iteration, int chan_idx, int bytes);
+
+void send_sdu_chan_worker(struct k_work *item)
+{
+	const struct channel *ch = CONTAINER_OF(item, struct channel, work);
+
+	send_sdu(ch->iteration, ch->chan_id, ch->bytes_to_send);
+}
+
+static void send_sdu(int iteration, int chan_idx, int bytes)
+{
+	struct bt_l2cap_chan *chan = &channels[chan_idx].le.chan;
+	struct net_buf *buf;
+
+	*((int *)channels[chan_idx].payload) =
+		iteration; /*First 4 bytes in sent payload is iteration count*/
+	if (channels[chan_idx].buf != 0) {
+		FAIL("Buf should have been dealocated by now");
+	}
+
+	if (chan_idx == 0)
+		buf = net_buf_alloc(&tx_data_pool_0, K_NO_WAIT);
+	else
+		buf = net_buf_alloc(&tx_data_pool_1, K_NO_WAIT);
+
+	if (buf == NULL) {
+		FAIL("Failed to get buff on ch %i, iteration %i should never happen", chan_idx,
+		     chan_idx);
+	}
+
+	channels[chan_idx].buf = buf;
+	net_buf_reserve(buf, BT_L2CAP_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, channels[chan_idx].payload, bytes);
+
+	LOG_DBG("bt_l2cap_chan_sending ch: %i bytes: %i iteration: %i", chan_idx, bytes, iteration);
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	LOG_DBG("bt_l2cap_chan_send returned: %i", ret);
+
+	if (ret < 0) {
+		FAIL("Error: send failed error: %i", ret);
+		channels[chan_idx].buf = 0;
+		net_buf_unref(buf);
+	}
+}
+
+static void send_sdu_concurrently(void)
+{
+	for (int i = 0; i < SDU_SEND_COUNT; i++) {
+		for (int k = 0; k < L2CAP_CHANNELS; k++) {
+			channels[k].iteration = i;
+			channels[k].bytes_to_send = (k == 0) ? LONG_MSG : SHORT_MSG; /* assign less bytes to be sent for the channel 1 */
+			int err = k_work_submit_to_queue(&channels[k].work_queue, &channels[k].work);
+
+			if (err < 0){
+				FAIL("Failed to submit work to the queque, error: %d", err);
+			}
+		}
+		k_sem_take(&sent_sem, K_FOREVER);
```

Comment:
Given that we send twice (once per channel), the `sent` callback will be called twice. This `take` should thus be taken `L2CAP_CHANNELS` number of times

Reply:
Link doesn't work (did you do a rebase?)

Reply:
Ah, found it myself. 

It's different than how I would have implemented it, but it should do the job just as fine (personally I would have utilized the fact that semaphores have a count, and simply use that count when taking it here.